include/domain_decomposition.h

```diff
@@ -0,0 +1,153 @@
+      }
+    }
+
+    template <typename... Args> constexpr void set(const DD &flag, const Args &...args)
+    {
+      set(flag);
+      set(args...);
+    }
+
+    // Pretty print the args struct
+    void print() const
+    {
+      if (not *this) {
+        printfQuda("DD not in use\n");
+        return;
+      }
+      printfQuda("Printing DDParam\n");
+      for (int i = 0; i < (int)DD::size; i++)
+        printfQuda("flags[DD::%s] = %s\n", to_string((DD)i).c_str(), flags[i] ? "true" : "false");
+      for (int i = 0; i < QUDA_MAX_DIM; i++) printfQuda("block_dim[%d] = %d\n", i, static_cast<int>(block_dim[i]));
+    }
+
+    // Checks if this matches to given DDParam
+    template <typename F> inline bool check(const F &field, bool verbose = false) const
+    {
+      if (not *this) return true;
+
+      if (type == QUDA_DD_RED_BLACK) {
+        for (int i = 0; i < field.Ndim(); i++) {
+          if (block_dim[i] < 0) {
+            if (verbose) printfQuda("block_dim[%d] = %d is negative\n", i, block_dim[i]);
+            return false;
+          }
+          if (block_dim[i] > 0) {
+            int globalDim = comm_dim(i) * field.full_dim(i);
+            if (globalDim % block_dim[i] != 0) {
+              if (verbose) printfQuda("block_dim[%d] = %d does not divide %d \n", i, block_dim[i], globalDim);
+              return false;
+            }
+            if ((globalDim / block_dim[i]) % 2 != 0) {
+              if (verbose)
+                printfQuda("block_dim[%d] = %d does not divide %d **evenly** \n", i, block_dim[i], globalDim);
+              return false;
+            }
+          }
+        }
+        if (block_dim[0] % 2) {
+          if (verbose) printfQuda("block_dim[0] = %d must be even \n", block_dim[0]);
+          return false;
+        }
+      }
+
+      return true;
+    }
+
+    // Checks if this matches to given DDParam
+    inline bool match(const DDParam &dd, bool verbose = false) const
+    {
+      // if one of the two is not in use we return true, i.e. one of the two is a full field
+      if (not *this or not dd) return true;
+
+      // false if type does not match
+      if (type != dd.type) {
+        if (verbose) printfQuda("DD type do not match (%d != %d)\n", type, dd.type);
+        return false;
+      }
+
+      if (type == QUDA_DD_RED_BLACK) {
+        for (int i = 0; i < QUDA_MAX_DIM; i++)
+          if (block_dim[i] != dd.block_dim[i]) {
+            if (verbose) printfQuda("block_dim[%d] = %d != %d \n", i, block_dim[i], dd.block_dim[i]);
+            return false;
+          }
+        if (is(DD::no_block_hopping) != dd.is(DD::no_block_hopping)) {
+          if (verbose) printfQuda("no_block_hopping do not match.\n");
+          return false;
+        }
+      }
+
+      return true;
+    }
+
+    // Checks if this is equal to given DDParam
+    inline bool operator==(const DDParam &dd) const
+    {
+      // if both are not in use we return true
+      if (not *this and not dd) return true;
```

Comment:
Stylistic thing, we don't really use `not`, `and`, etc, elsewhere in QUDA, opting for the traditional `!`, `&&`, etc, instead. I think we should stick with the latter. Thoughts, @maddyscientist ? I'm willing to take the time to change the `not`/`and`/etc.

Reply:
I think I agree with @weinbe2 here, and we should stick to legacy logical operators here.

Reply:
I'll test and ideally create a PR to automagically handle this today, per @mathiaswagner 's suggestion offline. (Reference for the suggestion here: https://clang.llvm.org/extra/clang-tidy/checks/readability/operators-representation.html)